modules/dnn/include/opencv2/dnn/dnn.hpp

```diff
@@ -56,7 +56,35 @@ CV__DNN_INLINE_NS_BEGIN
 //! @addtogroup dnn
 //! @{
 
+    enum Layout
+    {
+        DNN_LAYOUT_UNKNOWN = 0,
+        DNN_LAYOUT_ND = 1,
+        DNN_LAYOUT_NCHW = 2,
+        DNN_LAYOUT_NHWC = 3,
+        DNN_LAYOUT_NCHWс = 4
+    };
+    
     typedef std::vector<int> MatShape;
+    struct CV_EXPORTS TensorShape
+    {
+        TensorShape();
+        template<typename _Tp> TensorShape(int ndims, const _Tp* shape,
+                                           int layout=DNN_LAYOUT_UNKNOWN);
+        template<typename _Tp> TensorShape(std::initializer_list<_Tp> shape,
+                                           int layout=DNN_LAYOUT_UNKNOWN);
+        static TensorShape fromArray(InputArray m, int ndims,
+                                     int layout=DNN_LAYOUT_UNKNOWN);
+        int toMatShape(int* mshape) const;
```

Comment:
Passing arrays without size is unsafe API.

Reply:
fixed